### PR TITLE
Support for metadata.generateName (CSA)

### LINKED ISSUE #2539: generate_name prefix seems to not be honored
### What happened?

I have both a `pv` and `pvc` being created, I specify the `pulumi_kubernetes.meta.v1.ObjectMetaArgs` `generate_name` attribute on each which should specify a prefix for the name.  But the name that results is not using my specified prefix, it appears to be using the `resource_name` of the object as the prefix instead.

### Expected Behavior

For example, my code for generating a `pvc` from the `pv` I created before
```
        # Create the PVC from the PV
        self.pvc = pulumi_kubernetes.core.v1.PersistentVolumeClaim(
            f'{self.cluster.name}-workbench-pvc',
            args=pulumi_kubernetes.core.v1.PersistentVolumeClaimInitArgs(
                metadata=pulumi_kubernetes.meta.v1.ObjectMetaArgs(
                    generate_name='workbench-pvc',
                    namespace='datascience'
                ),
                spec=pulumi_kubernetes.core.v1.PersistentVolumeClaimSpecArgs(
                    access_modes=['ReadWriteMany'],
                    resources=pulumi_kubernetes.core.v1.ResourceRequirementsArgs(
                        requests=capacity
                    ),
                    storage_class_name='',
                    volume_name=pv.metadata.name
                )
            ),
            opts=pulumi.ResourceOptions(provider=self.cluster.provider,
                                        parent=pv)
        )
```

I would expect the name of the `pvc` to be `workbench-pvc-<randomstring>` but instead I get `<cluster_name>-workbench-pvc-<randomstring>`

our cluster name is `internal-eks01` in this case
```
internal-eks01-workbench-pvc-c1aa4e6e         Bound    internal-eks01-workbench-pv-abae85d9         1200Gi     RWX                           55d
```

### Steps to reproduce

Create a `pvc` from a `pv` using the code below

 ```
        # Create the PVC from the PV
        self.pvc = pulumi_kubernetes.core.v1.PersistentVolumeClaim(
            f'{self.cluster.name}-workbench-pvc',
            args=pulumi_kubernetes.core.v1.PersistentVolumeClaimInitArgs(
                metadata=pulumi_kubernetes.meta.v1.ObjectMetaArgs(
                    generate_name='workbench-pvc',
                    namespace='datascience'
                ),
                spec=pulumi_kubernetes.core.v1.PersistentVolumeClaimSpecArgs(
                    access_modes=['ReadWriteMany'],
                    resources=pulumi_kubernetes.core.v1.ResourceRequirementsArgs(
                        requests=capacity
                    ),
                    storage_class_name='',
                    volume_name=pv.metadata.name
                )
            ),
            opts=pulumi.ResourceOptions(provider=self.cluster.provider,
                                        parent=pv)
        )
```

it will not use the `generate_name` prefix, it instead uses the `resource_name` as a prefix

### Output of `pulumi about`

```
❯ pulumi about
CLI
Version      3.77.1
Go Version   go1.20.7
Go Compiler  gc

Plugins
NAME                     VERSION
aws                      5.41.0
aws-native               0.72.0
kubernetes               3.28.1
kubernetes-cert-manager  0.0.5
postgresql               3.8.0
python                   unknown

Host
OS       darwin
Version  12.6.8
Arch     arm64

This project is written in python: executable='/Users/jforest/.pyenv/shims/python3' version='3.9.11
'

<SNIP SOME RESOURCES OUT>

Found no pending operations associated with jforest

Current Stack: jforest/aws-main-services/jforest
Backend
Name           pulumi.com
URL            https://app.pulumi.com/jforest
User           jforest
Organizations  jforest
```

### Additional context

versions of pulumi libraries:

```
❯ pip list | grep pulumi
pulumi                         3.68.0
pulumi-aws                     5.41.0
pulumi-aws-native              0.72.0
pulumi-kubernetes              3.28.1
pulumi-kubernetes-cert-manager 0.0.5
pulumi-postgresql              3.8.0
```

I would like to upgrade `pulumi-kubernetes` to 4.1.0, but `pulumi-kubernetes-cert-manager` will not allow versions greater than 3.x.  https://github.com/pulumi/pulumi-kubernetes-cert-manager/issues/21 and https://github.com/pulumi/pulumi-kubernetes-cert-manager/pull/22

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by EronWright**
### Proposed changes

This PR implements support for `.metadata.generateName` in CSA mode, based on https://github.com/pulumi/pulumi-kubernetes/pull/2790.
1. Adjust the auto-naming logic to consider `.metadata.generateName` to be a variant of auto-naming. Pulumi will not assign an auto-name to a new resource that has `generateName`, and upon delete will use the replace-then-delete technique.
2. Use outputs rather than inputs to fetch the live object or to obtain the live object's name.
3. Fix an autonaming bug where the old auto-name would erroneously be adopted, overwriting a new computed name. 

### Tests
1. A new integration test is provided to test the use of `metadata.generateName`.  It tests creation, update, replacement, and promotion from `.generateName` to `.name`.  ([ref](https://github.com/pulumi/pulumi-kubernetes/pull/2808/files#diff-7e85ce707283b1d281d971a8e5b0c49b959b0803ba3437dc9dbd422552326835R269))
2. New test cases for the await logic.
3. The existing autonaming test is enhanced to test how autonaming takes precedence over `generateName`, in the update case. This is to ensure backwards compatibility, e.g. in the edge case that an existing object has a `generateName` field (which Pulumi ignored until now).

### Related issues (optional)
Closes #2539

### CODE REVIEW

File: provider/pkg/provider/provider.go [hunk 1]
```diff
@@ -1552,6 +1559,9 @@ func (k *kubeProvider) Diff(ctx context.Context, req *pulumirpc.DiffRequest) (*p
 	newInputs := propMapToUnstructured(newResInputs)
 
 	oldInputs, oldLive := parseCheckpointObject(oldState)
+	if !isHelmRelease(urn) {
+		contract.Assertf(oldLive.GetName() != "", "expected live object name to be nonempty: %v", oldLive)
+	}
```
Comment (EronWright) [inline provider/pkg/provider/provider.go:1564]: This is a workaround for https://github.com/pulumi/pulumi-kubernetes/issues/2679 and I'll add a comment before merging.

File: provider/pkg/provider/provider.go [hunk 2]
```diff
@@ -1600,6 +1611,12 @@ func (k *kubeProvider) Diff(ctx context.Context, req *pulumirpc.DiffRequest) (*p
 	if k.serverSideApplyMode && len(oldLivePruned.GetResourceVersion()) > 0 {
 		oldLivePruned.SetResourceVersion("")
 	}
+	// If a name was specified in the new inputs, be sure that the old live object has the previous name.
+	// This makes it possible to update the program to set `.metadata.name` to the name that was
+	// made by `.metadata.generateName` without triggering replacement.
+	if newInputs.GetName() != "" {
+		oldLivePruned.SetName(oldLive.GetName())
```
Comment (rquitales) [inline provider/pkg/provider/provider.go:1618]: Admittedly, this took me a couple of passes to understand what's happening. Maybe it's worthwhile to expand the comment a bit? Something like, generateName will create and store a name, but the oldLivePruned object drops the name field to match the input, which we now need for diffing.

Though this could just be an artifact of code review only showing small chunks of code without context.

File: tests/sdk/nodejs/autonaming/step4/index.ts [hunk 3]
```diff
@@ -14,17 +14,17 @@
 
 import * as k8s from "@pulumi/kubernetes";
 
-export const namespace = new k8s.core.v1.Namespace("test-namespace");
+const namespace = new k8s.core.v1.Namespace("test-namespace");
 
 //
-// User has now specified `.metadata.name`, so Pulumi should replace the resource, and NOT allocate
-// a name to it.
+// User has now specified `.metadata.generateName`, which Pulumi ignores because autonaming has already occurred,
```
Comment (rquitales) [inline tests/sdk/nodejs/autonaming/step4/index.ts:20]: In this scenario, could we emit a warning to the user that we won't honor `generateName`? I believe we just silently continue right?

File: tests/sdk/nodejs/generatename/step2/index.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+import * as pulumi from "@pulumi/pulumi";
+import * as k8s from "@pulumi/kubernetes";
+
+const config = new pulumi.Config();
+
+const namespace = new k8s.core.v1.Namespace("test-namespace");
+
+//
+// The `.metadata.generateName` field has changed, but Pulumi does NOT automatically replace in that situation.
```
Comment (rquitales) [inline tests/sdk/nodejs/generatename/step2/index.ts:23]: Maybe a warning here too?

**Review by rquitales — APPROVED** [review]
LGTM

**Comment by EronWright** [comment]
@vavsab I would guess that your provider is configured to use server-side apply mode, which you can disable:
https://www.pulumi.com/registry/packages/kubernetes/how-to-guides/managing-resources-with-server-side-apply/

**Comment by gunzy83** [comment]
> @vavsab I would guess that your provider is configured to use server-side apply mode, which you can disable: https://www.pulumi.com/registry/packages/kubernetes/how-to-guides/managing-resources-with-server-side-apply/

Is there a reason this is not supported with SSA? Will it be supported in the future?

**Comment by EronWright** [comment]
The reason that it isn't supported is that, in SSA mode, the provider uses Apply (HTTP PATCH) to both create and update the object (as described [here](https://kubernetes.io/docs/reference/using-api/api-concepts/#patch-and-apply)), and one cannot use `.metadata.generateName` in that case. See https://github.com/pulumi/pulumi-kubernetes/issues/2926 for a discussion about the behavior should be changed to use Create (HTTP POST) even in SSA mode, which would have the nice side-effect of enabling this feature.

**Comment by gunzy83** [comment]
> The reason that it isn't supported is that, in SSA mode, the provider uses Apply (HTTP PATCH) to both create and update the object (as described [here](https://kubernetes.io/docs/reference/using-api/api-concepts/#patch-and-apply)), and one cannot use `.metadata.generateName` in that case. See #2926 for a discussion on whether the behavior should be changed to use Create (HTTP POST) even in SSA mode, which would have the nice side-effect of enabling this feature.

Thanks, I understand now. I will follow the issue over there.
